compiler/rustc_mir_transform/src/generator.rs

```diff
@@ -331,22 +335,44 @@ impl<'tcx> MutVisitor<'tcx> for TransformVisitor<'tcx> {
         });
 
         let ret_val = match data.terminator().kind {
-            TerminatorKind::Return => Some((
-                VariantIdx::new(1),
-                None,
-                Operand::Move(Place::from(self.new_ret_local)),
-                None,
-            )),
+            TerminatorKind::Return => {
+                Some((true, None, Operand::Move(Place::from(self.new_ret_local)), None))
+            }
             TerminatorKind::Yield { ref value, resume, resume_arg, drop } => {
-                Some((VariantIdx::new(0), Some((resume, resume_arg)), value.clone(), drop))
+                Some((false, Some((resume, resume_arg)), value.clone(), drop))
             }
             _ => None,
         };
 
-        if let Some((state_idx, resume, v, drop)) = ret_val {
+        if let Some((is_return, resume, v, drop)) = ret_val {
             let source_info = data.terminator().source_info;
             // We must assign the value first in case it gets declared dead below
-            data.statements.extend(self.make_state(state_idx, v, source_info));
+            let state_idx = VariantIdx::new(match (is_return, self.is_async_kind) {
+                (true, false) => 1,  // GeneratorState::Complete
+                (false, false) => 0, // GeneratorState::Yielded
+                (true, true) => 0,   // Poll::Ready
+                (false, true) => 1,  // Poll::Pending
+            });
+            let (kind, ty) = self.make_state(state_idx);
+            if let Some(ty) = ty {
+                data.statements.extend(expand_aggregate(
+                    Place::return_place(),
+                    std::iter::once((v, ty)),
+                    kind,
+                    source_info,
+                    self.tcx,
+                ));
+            } else {
+                // FIXME(swatinem): assert that `val` is indeed unit?
+                data.statements.extend(expand_aggregate(
+                    Place::return_place(),
+                    std::iter::empty(),
+                    kind,
+                    source_info,
+                    self.tcx,
+                ));
+            }
```

Comment:
It would be more readable if this block went back into `make_state`, eventually taking an `&mut Vec<Statement>`.